src/textual/worker.py

```diff
@@ -0,0 +1,388 @@
+            assert callable(self._work)
+            loop = asyncio.get_running_loop()
+
+            def run_work(work: Callable[[], ResultType]) -> ResultType:
+                """Set the active worker, and run the work."""
+                active_worker.set(self)
+                return work()
+
+            result = await loop.run_in_executor(None, run_work, self._work)
+        return result
+
+    async def _run(self, app: App) -> None:
+        """Run the worker.
+
+        Args:
+            app: App instance.
+        """
+        app._set_active()
+        active_worker.set(self)
+
+        self.state = WorkerState.RUNNING
+        app.log.worker(self)
+        try:
+            self._result = await self.run()
+        except asyncio.CancelledError as error:
+            self.state = WorkerState.CANCELLED
+            self._error = error
+            app.log.worker(self)
+        except Exception as error:
+            self.state = WorkerState.ERROR
+            self._error = error
+            app.log.worker(self, "failed", repr(error))
+            app.log.worker(Traceback())
+            if self.exit_on_error:
+                app.fatal_error()
+        else:
+            self.state = WorkerState.SUCCESS
+            app.log.worker(self)
+
+    def _start(
+        self, app: App, done_callback: Callable[[Worker], None] | None = None
+    ) -> None:
+        """Start the worker.
+
+        Args:
+            app: An app instance.
+            done_callback: A callback to call when the task is done.
+        """
+        if self._task is not None:
+            return
+        self.state = WorkerState.RUNNING
+        self._task = asyncio.create_task(self._run(app))
+
+        def task_done_callback(_task: asyncio.Task) -> None:
+            """Run the callback.
+
+            Called by `Task.add_done_callback`.
+
+            Args:
+                The worker's task.
+            """
+            if done_callback is not None:
+                done_callback(self)
+
+        self._task.add_done_callback(task_done_callback)
+
+    def cancel(self) -> None:
+        """Cancel the task."""
+        self._cancelled = True
+        if self._task is not None:
+            self._task.cancel()
+
+    async def wait(self) -> ResultType:
+        """Wait for the work to complete.
+
+        Raises:
+            WorkerFailed: If the Worker raised an exception
+            WorkerCancelled: If the Worker was cancelled before it completed.
+
+        Returns:
+            The return value of the work.
+        """
+        try:
+            if active_worker.get() is self:
+                raise DeadlockError(
+                    "Can't call worker.wait from within the worker function!"
+                )
+        except LookupError:
+            # Not in a worker
+            pass
+
+        if self.state == WorkerState.PENDING:
+            raise WorkerError("Worker must be started before calling this method.")
+        if self._task is not None:
```

Comment:
Shouldn't we also check that `self.state` is `RUNNING` or if I await a worker that has been completed, it won't run again?